src/agent/src/device.rs

```diff
@@ -225,12 +223,13 @@ struct VirtioBlkPciMatcher {
 }
 
 impl VirtioBlkPciMatcher {
-    fn new(relpath: &str) -> VirtioBlkPciMatcher {
+    fn new(relpath: &str) -> Result<VirtioBlkPciMatcher> {
         let root_bus = create_pci_root_bus_path();
         let re = format!(r"^{}{}/virtio[0-9]+/block/", root_bus, relpath);
-        VirtioBlkPciMatcher {
-            rex: Regex::new(&re).unwrap(),
-        }
+
+        Ok(VirtioBlkPciMatcher {
+            rex: Regex::new(&re)?,
+        })
```

Comment:
I understand exactly what you're saying, but:

- API: If there was no point in checking the value from `Regex::new()`, it wouldn't return an error.
- Maintenance: Although the code is *currently* correct, who knows what changes that code will under go in the future, so I'd always vote for the most cautious / paranoid / defensive strategy.
- `?` is shorter and less visually distracting than `.unwrap()` :smile:

Reply:
Well, I don't feel strongly enough about this to insist on this point, but I do disagree here:

> I understand exactly what you're saying, but:
> 
>     * API: If there was no point in checking the value from `Regex::new()`, it wouldn't return an error.

The API makes sense for `Regex::new` because it could be getting user data, so needs to validate and report it.  But in our case the regex is a literal, so a failure here is a bug in our code not a situational problem.

>     * Maintenance: Although the code is _currently_ correct, who knows what changes that code will under go in the future, so I'd always vote for the most cautious / paranoid / defensive strategy.

Theoretically, yes, but this is not fragile against likely code changes.  In any situation I can imagine, it would be a code bug for us to generate an invalid regex string here.

>     * `?` is shorter and less visually distracting than `.unwrap()` smile

Well, sure.

My counterpoint is that as a rule problems that are necessarily bugs in our code should be reported as asserts or panics, not as error codes.  It clearly distinguishes a "this should not happen, ever" case from "maybe my configuration is bad", and it puts the crash as close to the bug as we can, which helps with debugging.

Admittedly this is offset in the specific case of Kata, since the way we package it means that agent panics aren't properly reported up to somewhere useful without jumping through a lot of hoops :(.

Reply:
Ok. I've removed the `Result` return value from `VirtioBlkPciMatcher` and changed the original `unwrap()` to an `expect`()` for `VirtioBlkPciMatcher` and `VirtioBlkCCWMatcher` so we get a slightly better message that shows this is an internal bug.

The remaining `Regex::new()` shouldn't use `unwrap()` though since the string passed to that call is coming via a parameter (aka untrusted source).